src/compiler/moduleNameResolver.ts

```diff
@@ -67,40 +67,32 @@ namespace ts {
     }
 
     /** Reads from "main" or "types"/"typings" depending on `extensions`. */
-    function tryReadPackageJsonMainOrTypes(extensions: Extensions, packageJsonPath: string, baseDirectory: string, state: ModuleResolutionState): string {
+    function tryReadPackageJsonFields(ts: boolean, packageJsonPath: string, baseDirectory: string, state: ModuleResolutionState): string | undefined {
         const jsonContent = readJson(packageJsonPath, state.host);
+        const file = ts ? tryReadFromField("typings") || tryReadFromField("types") : tryReadFromField("main");
+        if (!file && state.traceEnabled) {
+            trace(state.host, Diagnostics.package_json_does_not_have_a_0_field, ts ? "types" : "main");

```

Comment:
it looks that we'll get this error even in case if `package.json` has field with some valid name but incorrect type of the value (`tryReadFromField` in this case will return `undefined`) - it is a bit misleading